### ISSUE TITLE
Interface improvement suggestions for tf.tranpose and tf.slice

### ISSUE THREAD

**Issue by rizar** [tensorflow/tensorflow#4516]
I would like to suggest the following improvements for the interfaces of `tf.transpose` and `tf.slice`.
1. `tf.transpose` currently accepts a boolean mask that selects the dimensions to be reversed. AFAIK this is the only op in Tensorflow that uses a mask for this purpose, whereas other, e.g. `tf.reduce_***`, take an axis number or a list of axis numbers. I think `tf.tranpose` could also be changed accordingly.
2. `tf.slice` signature is `tf.slice(input_, begin, size, name=None)`, and it is very unusual for Python that the slices are defined by their size, not by their end. Also, there is a `slice` object in Python, which would be a very natural argument for `tf.slice`.

**Comment by yaroslavvb** [comment]
BTW, there's tf.strided_slice
<https://www.tensorflow.org/versions/r0.10/api_docs/python/array_ops.html#strided_slice>
which is closer to the API you want.
The issue with changing tf.slice is that a lot of models will get broken by
this change

PS, here's the function that is used by TensorFlow to turn Python slice
objects into slice specifications
https://github.com/tensorflow/tensorflow/blob/640353d5d1db50d6601f9410b9d06462a8a71ce4/tensorflow/python/ops/array_ops.py#L262

On Wed, Sep 21, 2016 at 11:40 AM, Dzmitry Bahdanau <notifications@github.com

**Comment by rizar** [comment]
> The issue with changing tf.slice is that a lot of models will get broken by
this change

Should it be deprecated then, if it's only purpose is backward compatibility?

**Comment by concretevitamin** [comment]
@aselle Is deprecating `tf.slice()` on your list?

**Comment by aselle** [comment]
I have no immediate plans to deprecate tf.slice(). I think we should update the documentation, but I'm not sure we should deprecate it even in 1.0.

**Comment by rizar** [comment]
But what's the point of having both tf.slice and tf.strided_slice? Is there
anything that the latter can not do and the former can?

**Comment by yaroslavvb** [comment]
@rizar removing tf.slice would break existing models which use it

**Comment by rizar** [comment]
I understand that. But I thought that this is exactly what deprecation is
used for. By deprecating tf.slice officially you could encourage people to
not use it in new projects and replace it with tf.strided_slice in the old
ones. This way it could be eventually removed.

**Comment by aselle** [comment]
I of course would like to get rid of redundancy, however there is always a tradeoff. While it is true that deprecation helps make changes more graceful, it does not make them zero cost for users. It just defers that cost to a potentially more convenient time in the future instead of forcing it immediately. We are thinking about deprecating it, but I don't know when that will happen.

**Comment by aselle** [comment]
In any case, back to the original issue. 
1. tf.slice: I will add a pointer to strided_slice in the slice documentation for now as it works like you are suggesting.
2. tf.transpose: I don't know what you mean by mask. You specify which axis indices should be swapped i.e. for a rank-2 matrix  A, tf.transpose(A, [1,0]) is normal transposing and tf.transpose(A,[0,1]) is identity. This is consistent with how NumPy works.

Are you ok with me closing the issue?

**Comment by rizar** [comment]
1. Adding a pointer would alleviate the issue, thanks.
2. This is totally my bad: I confused `tf.transpose` and `tf.reverse`. It's the interface of the latter that I think could be improved. See the documentation below:

```
Reverses specific dimensions of a tensor.

Given a `tensor`, and a `bool` tensor `dims` representing the dimensions
of `tensor`, this operation reverses each dimension i of `tensor` where
`dims[i]` is `True`.
```

The interface of `tf.reverse` stands out from interfaces of other tf functions. such as e.g. `tf.transpose`, by accepting a boolean mask of dimensions instead of a list of dimensions.

**Comment by aselle** [comment]
The `tf.reverse` part of the issue has been resolved as of TensorFlow 1.0, and I just resolved the documentation issue (pointing users from tf.slice to tf.getitem).
